ratatui-widgets/src/table.rs

```diff
@@ -2276,4 +2315,157 @@ mod tests {
         let column_count = table.column_count();
         assert_eq!(column_count, expected);
     }
+
+    #[rstest]
+    #[case::no_padding(4, 2, 0, Some(2), Buffer::with_lines(vec![">> 2 ", "   3 ", "   4 ", "   5 "]))]
+    #[case::one_before(
+        4,
+        2, // Offset
+        1, // Padding
+        Some(2), // Selected
+        Buffer::with_lines(vec!["   1 ", ">> 2 ", "   3 ", "   4 "])
+    )]
+    #[case::check_padding_overflow(
+        4,
+        1, // Offset
+        2, // Padding
+        Some(4), // Selected
+        Buffer::with_lines(vec!["   2 ", "   3 ", ">> 4 ", "   5 "])
+    )]
+    #[case::no_padding_offset_behavior(
+        5, // Render Area Height
+        2, // Offset
+        0, // Padding
+        Some(3), // Selected
+        Buffer::with_lines(
+            vec!["   2 ", ">> 3 ", "   4 ", "   5 ", "   6 "]
+            )
+    )]
+    #[case::two_before(
+        5, // Render Area Height
+        2, // Offset
+        2, // Padding
+        Some(3), // Selected
+        Buffer::with_lines(
+            vec!["   1 ", "   2 ", ">> 3 ", "   4 ", "   5 "]
+            )
+    )]
+    #[case::keep_selected_visible(
+        4,
+        0, // Offset
+        4, // Padding
+        Some(1), // Selected
+        Buffer::with_lines(vec!["   0 ", ">> 1 ", "   2 ", "   3 "])
+    )]
+    fn test_padding(
+        #[case] render_height: u16,
+        #[case] offset: usize,
+        #[case] padding: usize,
+        #[case] selected: Option<usize>,
+        #[case] expected: Buffer,
+    ) {
+        let rows = (0..8).map(|i| Row::new([i.to_string()]));
+        let table = Table::new(rows, [Constraint::Length(2)])
+            .scroll_padding(padding)
+            .highlight_symbol(">> ");
+        let mut buf = Buffer::empty(Rect::new(0, 0, 5, render_height));
+        let mut state = TableState::new()
+            .with_offset(offset)
+            .with_selected(selected);
+
+        StatefulWidget::render(
+            table.clone(),
+            Rect::new(0, 0, 5, render_height),
+            &mut buf,
+            &mut state,
+        );
+
+        assert_eq!(expected, buf);
+    }
+
+    /// If there isn't enough room for the selected item and the requested padding the table can
```

Comment:
Is this a current bug? If there's an issue for it already, link to it to help tell the full story. Otherwise can you explain the problem in a bit more detail? I'm not sure that I fully understand the problem and it's not obvious which parts of the test are part of it.

It's good to have tests like this called out as regression tests to prevent bugs that we've fixed happening again, so I definitely appreciate a good comment like this. Thanks!

One way of showing the problem is to write a characterization test like this before the fix in one commit and then show how the fix changes the test. (Alternatively leave a comment somewhere in the test that shows that you've run the test with the code prior to the fix and note the previous value which changed).